lib/rialto/etl/cli/grants.rb

```diff
@@ -0,0 +1,80 @@
+# frozen_string_literal: true
+
+require 'fileutils'
+require 'parallel'
+
+module Rialto
+  module Etl
+    module CLI
+      # Grants iterates over the people in the input file JSON and for each
+      # valid UID (SUNet ID) calls three other commands:
+      #   exe/extract call Sera --sunetid username > username.json
+      #   TODO: exe/transform call Sera -i username.json > username.sparql
+      #   TODO: exe/load call Sparql -i username.sparql
+      class Grants < Thor
+        option :input_file,
+               required: true,
+               banner: 'FILENAME',
+               desc: 'Name of file with data to be loaded (REQUIRED)',
+               aliases: '-i'
+        option :dir,
+               required: false,
+               default: 'data',
+               banner: 'DIR',
+               desc: 'Name of the directory to store data in',
+               aliases: '-d'
+        option :batch_size,
+               required: false,
+               default: 3,
+               banner: 'BATCH_SIZE',
+               desc: 'Size of batch for parallel processing',
+               aliases: '-s'
+        option :force,
+               required: false,
+               default: false,
+               banner: 'FORCE',
+               desc: 'Overwrite files that already exist',
+               aliases: '-f'
+
+        desc 'Load', 'Load all grants for all researchers in the file'
+        # rubocop:disable Metrics/MethodLength
+        # rubocop:disable Metrics/AbcSize
+        def load
+          FileUtils.mkdir_p(options[:dir])
+          File.open(options[:input_file], 'r') do |f|
+            f.each_line.lazy.each_slice(options[:batch_size].to_i) do |lines|
+              batch_ids = lines.map { |line| JSON.parse(line)['sunetid'] }
+              Parallel.map(batch_ids, in_processes: options[:batch_size].to_i) do |id|
+                output_file = File.join(options[:dir], "#{id}.json")
+                if File.exist?(output_file) && !options[:force]
+                  say "file #{output_file} already exists, skipping. use -f to force overwrite"
+                  next
+                end
+                extract_and_write(id, output_file)
+              end
+            end
+          end
+        end
+        # rubocop:enable Metrics/AbcSize
+
+        private
+
+        def extract_and_write(id, output_file)
+          say "extracting SeRA records for #{id}"
+          results = []
+          Rialto::Etl::Extractors::Sera.new(sunetid: id).each do |result|
+            results << result
+          end
+          return if results.empty?
+          File.open(output_file, 'w') do |f|
+            f.write(results.join("\n"))
+          end
+        rescue StandardError => exception
+          say "retrying request for #{id}, failed: #{exception.message}"
+          retry
```

Comment:
So this will retry if any exception is thrown, including a file I/O related problem?

For a SeRA API related exception/error - what do you think about waiting a little before retrying? Just in the extremely unlikely event this highly robust API should have a hiccup one day..?

Reply:
Yes, this will retry if any exception is thrown. One reason for that at the moment is I don't know in practice what errors are thrown. I'm hoping to see some fails in the current run so I can refine this `rescue`. Alternatively, or in addition, we could consider retrying 3-5 times only so we don't retry *ad infinitum*?

Reply:
The three kinds of errors I see so far look like:

```
There was a problem with the request to `/mais/sera/v1/api?scope=sera.stanford-only&sunetId=username` (404) # or (500)
```

```
retrying request for username, failed: Connection reset by peer - SSL_connect
```

```
Failed to open TCP connection to aswsuat.stanford.edu:443 (getaddrinfo: Temporary failure in name resolution)
```

To start, I could grep around to figure out what exception class the bottom two raise. (The top is a `StandardError` because we raise it ourselves -- we could also consider defining a custom exception class and raising that instead, to be more fine-grained.)

Reply:
I've stopped the harvest to add the exception class to the error message, and then restarted it. I believe the second exception is `Errno::ECONNRESET`, and the third exception is `SocketError`.

Reply:
Check out the changes I've made to this file re: exception handling.